### PR TITLE
feat: Summarisation of user reviews #Issue17

### PR DESCRIPTION
**Pull request by rajput-kshitij**
PR checklist:

- [x] Github Issue associated with branch.
- [x] Code passes linting through autopep8.
- [x] PR branch is up-to-date with latest master branch.

### CODE REVIEW

File: app/sample-mint-config.json [hunk 1]
```diff
@@ -38,6 +38,9 @@
         "story_type": "12",
         "bug_type": "1"
     },
+    "summary_config": {
```
Comment (MadaraUchiha-314) [inline app/sample-mint-config.json:41]: This shoudl come under algorithm config

File: fawkes/algorithms/algo.py [hunk 2]
```diff
@@ -23,6 +23,55 @@
 from fawkes.review.review import Review
 from fawkes.cli.fawkes_actions import FawkesActions
 
+from sentence_transformers import SentenceTransformer
+import numpy as np
+from sklearn.cluster import KMeans
+from gensim.summarization.summarizer import summarize
+
+
+'''
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/algo.py:32]: Double quotes pls. Its the standard followed throughout the repo.

File: fawkes/constants/constants.py [hunk 3]
```diff
@@ -123,3 +126,11 @@
...
 WEEKLY_EMAIL_DETAILED_TEMPLATE = "fawkes/email_summary/templates/weekly-detailed.html"
 WEEKLY_EMAIL_DETAILED_REVIEW_BLOCK_TEMPLATE = "fawkes/email_summary/templates/weekly-detailed-review-block.html"
+
+#Summrization constants
+num_clusters = 2
+# num_result_per_review=2
+minimum_reviews_per_cluster=10 # OR based on similarity in a cluster ( only retun in high rating)
+max_words_in_sentence= 20
+min_words_in_sentence= 4
+summary_length_per_cluster=30
```
Comment (MadaraUchiha-314) [inline fawkes/constants/constants.py:136]: Why are all these configs defined here ? Can these be user-configurable ?

File: fawkes/algorithms/algo.py [hunk 4]
```diff
@@ -23,6 +23,55 @@
...
+        @param{list<string>}: sentences - List of review sentence
+        @param{Integer}: num_clusters - User defined number of clusters
+        @returns{list<list<string>}: clustered_sentences - List of clusters of sentences 
+        
+        Groups similar sentences into clusters
+'''
+#Reference : https://github.com/UKPLab/sentence-transformers
+def k_means_classification(sentences, num_clusters):
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/algo.py:40]: Why can't all this logic be put in the summarisation file ?

Provide a clean API for the algorithm to use.

File: fawkes/algorithms/algo.py [hunk 5]
```diff
@@ -23,6 +23,55 @@
...
+    return clustered_sentences
+
+'''
+        @param{string}: text - text corpus to summarize
+        @param{integer}: word_count - max number of words in a summary
+        @returns{list<string>}: generated summary
+        
+        function to summarize a piece of text
+'''
+def summarize_text(text,word_count):
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/algo.py:68]: If summarisation is an algorithm, then the module should not write to a file on its own.

What algorithm does to modify the derived insights object for a Review.

I think you guys have mixed both the concepts.

File: fawkes/algorithms/summarization/summarization.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                clustered_sentences = k_means_classification(sentences, app_config.summary_config.num_clusters)
+                for i in range(len(clustered_sentences)):
+                    cluster = clustered_sentences[i]
+                    if(len(cluster) < constants.minimum_reviews_per_cluster):
+                        continue
+                    text = ". ".join(cluster) 
+                    gen_summary = summarize_text(text,constants.summary_length_per_cluster)
+                    summarized_category_review.append(gen_summary)
+            else:
+                print("Found very few sentences in category ="+ category)
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/summarization/summarization.py:102]: There is a clean way to do logging in fawkes. Check other files.

I agree that not all places have been refactored to use that, but all new code should use it.

**Review by MadaraUchiha-314 — CHANGES_REQUESTED** [review]
Can you also run the linter once ?

Lot of places where code style is not consistent with the overall project.

File: fawkes/constants/constants.py (hunk 3 above)
Comment (rohit256) [reply fawkes/constants/constants.py:136]: These are configs for the code, for eg. max_words_in_sentence= 20, min_words_in_sentence= 4 are specific to the code. The user should ideally have nothing to do with these

File: fawkes/algorithms/algo.py (hunk 4 above)
Comment (rohit256) [reply fawkes/algorithms/algo.py:40]: These are plugable algorithms, these might change based on performance. Why should the main code to summarise have these ?

File: fawkes/algorithms/algo.py (hunk 5 above)
Comment (rohit256) [reply fawkes/algorithms/algo.py:68]: Did not understand this, summarize_text in algo.py is not writing into the file. Could you re-explain

**Comment by rohit256** [comment]
Could you share how can we run a linter please. Couldn't find anything in contributing.md or wiki :/

File: fawkes/algorithms/algo.py (hunk 4 above)
Comment (MadaraUchiha-314) [reply fawkes/algorithms/algo.py:40]: Then these should be in a separate module. The algo.py should not be polluted with these code.

File: fawkes/algorithms/algo.py (hunk 5 above)
Comment (MadaraUchiha-314) [reply fawkes/algorithms/algo.py:68]: Why does summarize_text has to be algo.py ? If its not a pluggable algorithm that I can run in fawkes ?

Comment (rohit256) [reply fawkes/algorithms/algo.py:68]: Ohh okay, i interpreted it the wrong way then. Will make the changes, Thanks!

**Comment by rajput-kshitij** [comment]
@MadaraUchiha-314 Addressed review comments. Can you please take a look.

File: fawkes/algorithms/summarization/summarization.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            categorized_review = reviews[category]
+
+            # Preprocess reviews
+            sentences = preprocess_review(categorized_review)
+            # number of sentences in a category should be atleast greater than
+            # the number of clusters
+            if(len(sentences) > app_config.algorithm_config.summarization.num_clusters - 1):
+                clustered_sentences = k_means_classification(
+                    sentences, app_config.algorithm_config.summarization.num_clusters)
+                for i in range(len(clustered_sentences)):
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/summarization/summarization.py:152]: If you look at this, there is no use of the "index". Can you use a for...in...loop over the elements, which is much cleaner ?

**Review by MadaraUchiha-314 — APPROVED** [review]

File: fawkes/algorithms/summarization/summarization.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+        # reject sentences of inappropriate lengths
+        for sentence in sentences:
+            word_count = len(sentence.split())
+            if(word_count > constants.min_words_in_sentence and word_count < constants.max_words_in_sentence):
+                processed_sentences.append(sentence)
+    return processed_sentences
+
+
+"""
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/summarization/summarization.py:98]: Documentation of function is written inside the function according to the standards and is whats followed in this repo.

File: fawkes/algorithms/summarization/summarization.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            app_name=app_config.app.name,
+        )
+
+        dir_name = os.path.dirname(query_results_file_path)
+        pathlib.Path(dir_name).mkdir(parents=True, exist_ok=True)
+
+        utils.dump_json(
+            [
+                {
+                    "review": summarized_reviews,
```
Comment (MadaraUchiha-314) [inline fawkes/algorithms/summarization/summarization.py:176]: Why is this key named "review" when there are multiple reviews in the list ?

**Review by MadaraUchiha-314 — APPROVED** [review]

**Pull request merged**